pitchtypes/spelled.py

```diff
@@ -0,0 +1,1067 @@
+        """
+        Comparison between two spelled types.
+
+        Returns 0 if the objects are equal,
+        1 if the first object (``self``) is greater,
+        and -1 if the second object (``other``) is greater.
+
+        The respective ordering differs between types.
+        Non-class pitches and intervals use diatonic ordering,
+        interval/pitch classes use line-of-fifths ordering.
+
+        This method can be indirectly used through binary comparison operators
+        (including ``==``, ``<`` etc.).
+
+        :param other: an object to compare to (same type as ``self``)
+        :return: ``-1`` / ``0`` / ``1`` (integer)
+        """
+        raise NotImplementedError
+
+    def __eq__(self, other):
+        return self.compare(other) == 0
+
+    def __lt__(self, other):
+        return self.compare(other) == -1
+    
+    def __le__(self, other):
+        return self.compare(other) != 1
+
+    # Spelled interface:
+
+    def fifths(self):
+        """
+        Return the position of the interval on the line of fifths.
+
+        :return: fifth position (integer)
+        """
+        raise NotImplementedError
+
+    def octaves(self):
+        """
+        For intervals, return the number of octaves the interval spans.
+        Negative intervals start with -1, decreasing.
+        For pitches, return the absolute octave of the pitch.
+
+        :return: external/independent octave (integer)
+        """
+        raise NotImplementedError
+
+    def internal_octaves(self):
+        """
+        Return the internal octave representation of a pitch,
+        which is dependent on the fifths.
+
+        Only use this if you know what you are doing.
+
+        :return: internal/dependent octave (integer)
+        """
+        raise NotImplementedError
+
+    def degree(self):
+        """
+        Return the "relative scale degree" (0-6) to which the interval points
+        (unison=0, 2nd=1, octave=0, 2nd down=6, etc.).
+        For pitches, return the integer that corresponds to the letter (C=0, D=1, ...).
+
+        :return: degree (integer)
+        """
+        return self._degree_from_fifths_(self.fifths())
+
+    def alteration(self):
+        """
+        Return the number of semitones by which the interval is altered from its the perfect or major variant.
+        Positive alteration always indicates augmentation,
+        negative alteration indicates diminution (minor or smaller) of the interval.
+        For pitches, return the accidentals (positive=sharps, negative=flats, 0=natural).
+
+        :return: alteration (integer)
+        """
+        raise NotImplementedError
+
+    def onehot(self):
+        """
+        Return a one-hot encoded tensor representing the object.
+        Specialized versions of this method take ranges for their respective dimensions.
+
+        :return: one-hot encoding of the object (numpy array)
+        """
+        raise NotImplementedError
+
+
+class SpelledI(abc.ABC):
```

Comment:
The naming is not very intuitive/self-explanatory, I would suggest to use `AbstractSpelledInterval` (and `AbstractSpelledPitch` below).

Reply:
Good point, those names are better.